### PR TITLE
feat: add pvc creation support

### PR DESCRIPTION
**Pull request by mt-czi**
Add the option for users to create pvc claims

### CODE REVIEW

File: stack/templates/_helpers.tpl [hunk 1]
```diff
@@ -103,3 +103,11 @@ image: {{ .Values.image }}
 image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
 {{- end }}
 {{- end }}
+
+{{- define "service.claimName" -}}
+{{- if and .Values.persistence.existingClaim }}
```
Comment (hspitzley-czi) [inline stack/templates/_helpers.tpl:108]:
```suggestion
{{- if .Values.persistence.existingClaim }}
```

File: stack/templates/deployment.yaml [hunk 2]
```diff
@@ -71,9 +71,15 @@ spec:
...
-          {{- with .Values.volumeMounts }}
+          {{- if or .Values.persistence.enabled .Values.volumeMounts}}
           volumeMounts:
-            {{- toYaml . | nindent 12 }}
+            {{- if .Values.persistence.enabled }}
+            - name: data
+              mountPath: {{ .Values.persistence.mountPath }}
+            {{- end }}
+            {{- if .Values.volumeMounts }}
+            {{- toYaml .Values.volumeMounts | nindent 8 }}
```
Comment (hspitzley-czi) [inline stack/templates/deployment.yaml:81]: Should `nindent` be changing from 12 to 8 here?

Comment (mt-czi) [reply stack/templates/deployment.yaml:81]: Yes, your correct! Just fixed it

File: stack/templates/deployment.yaml [hunk 3]
```diff
@@ -71,9 +71,15 @@ spec:
           {{- end }}
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
-          {{- with .Values.volumeMounts }}
+          {{- if or .Values.persistence.enabled .Values.volumeMounts}}
```
Comment (hspitzley-czi) [inline stack/templates/deployment.yaml:74]: Can you add defaults to `values.yaml` for the `persistence` fields? That's new, right?

Comment (mt-czi) [reply stack/templates/deployment.yaml:74]: Yes, I added them here: https://github.com/chanzuckerberg/argo-helm-charts/blob/GPU-271/pvc-creation-support/stack/values.yaml#L172

Comment (hspitzley-czi) [inline stack/templates/deployment.yaml:74]: I think `values.yaml` either needs a default mount path or this needs to check that the user sets a value for it
```suggestion
          {{- if or (and .Values.persistence.enabled .Values.persistence.mountPath) .Values.volumeMounts}}
```

File: stack/templates/deployment.yaml [hunk 4]
```diff
@@ -71,9 +71,15 @@ spec:
...
           volumeMounts:
-            {{- toYaml . | nindent 12 }}
+            {{- if .Values.persistence.enabled }}
```
Comment (hspitzley-czi) [inline stack/templates/deployment.yaml:76]:
```suggestion
            {{- if and .Values.persistence.enabled .Values.persistence.mountPath) }}
```

File: stack/templates/pvc.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  name: {{ include "service.fullname" . }}
+  labels:
+    {{- include "service.labels" . | nindent 4 }}
+  annotations:
+    {{- include "stack.annotations" $service | nindent 4 }}
+spec:
+  accessModes:
+  {{- range .Values.persistence.accessModes }}
+    - {{ . | quote }}
+  {{- end }}
```
Comment (hspitzley-czi) [inline stack/templates/pvc.yaml:23]: Can this be simplified to use `toYaml`?

Comment (jakeyheath) [reply stack/templates/pvc.yaml:23]: agreed, maybe just allow the user to specify a spec and toYaml it. That way we don't have to keep adding prs to support other features and the configuration looks more like natural k8s configuration

**Review by jakeyheath — APPROVED** [review]
looks good! thanks for your contribution!

small change to open up the PVC to allow all settings; also we don't have great testing and linting yet, so can you please do due diligence to make sure the helm template renders correctly and linted before merging.

File: stack/templates/deployment.yaml [hunk 6]
```diff
@@ -71,9 +71,15 @@ spec:
           {{- end }}
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
-          {{- with .Values.volumeMounts }}
+          {{- if or (and .Values.persistence.enabled .Values.persistence.mountPath) .Values.volumeMounts}}
           volumeMounts:
-            {{- toYaml . | nindent 12 }}
+            {{- if and .Values.persistence.enabled .Values.persistence.mountPath) }}
```
Comment (hspitzley-czi) [inline stack/templates/deployment.yaml:76]: Oops, my previous code suggestion wasn't quite right
```suggestion
            {{- if and .Values.persistence.enabled .Values.persistence.mountPath }}
```

**Comment by mt-czi** [comment]
@jakeyheath @hspitzley-czi Thank you for the feedback. I opened all the pvc settings and linted the chart

**Review by hspitzley-czi — APPROVED** [review]

**Pull request merged**